wandb/sdk/data_types/saved_model.py

```diff
@@ -0,0 +1,329 @@
+                )
+                json_obj["path"] = artifact.add_file(self._path, target_path, True).path
+        elif os.path.isdir(self._path):
+            # If the path is a directory, then we need to add all of the files
+            # The directory must be named deterministically based on the contents of the directory,
+            # but the files themselves need to have their name preserved.
+            # FUTURE: Add this functionality to the artifact adder itself
+            file_paths = []
+            for dirpath, _, filenames in os.walk(self._path, topdown=True):
+                for fn in filenames:
+                    file_paths.append(os.path.join(dirpath, fn))
+            dirname = b64_string_to_hex(md5_files_b64(file_paths))[:20]
+            target_path = os.path.join(".wb_data", "saved_models", dirname)
+            artifact.add_dir(self._path, target_path)
+            json_obj["path"] = target_path
+        else:
+            raise ValueError(
+                f"Expected a path to a file or directory, got {self._path}"
+            )
+
+        return json_obj
+
+    def model_obj(self) -> SavedModelObjType:
+        """ Returns the model object.
+        """
+        if self._model_obj is None:
+            assert self._path is not None, "Cannot load model object without path"
+            self._set_obj(self._deserialize(self._path))
+
+        assert self._model_obj is not None, "Model object is None"
+        return self._model_obj
+
+    # Methods to be implemented by subclasses
+    @staticmethod
+    def _deserialize(path: str) -> SavedModelObjType:
+        """Returns the model object from a path. Allowed to throw errors
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _validate_obj(obj: Any) -> bool:
+        """Validates the model object. Allowed to throw errors
+        """
+        raise NotImplementedError()
+
+    @staticmethod
+    def _serialize(obj: SavedModelObjType, dir_or_file_path: str) -> None:
+        """Save the model to disk. The method will receive a directory path which all
+        files needed for deserialization should be saved. A directory will always be passed if
+        _path_extension is an empty string, else a single file will be passed. Allowed to throw errors
+        """
+        raise NotImplementedError()
+
+    # Private Class Methods
+    @classmethod
+    def _maybe_init(
+        cls: Type["SavedModel"], obj_or_path: Any
+    ) -> Optional["SavedModel"]:
+        try:
+            return cls(obj_or_path)
+        except Exception as e:
+            if DEBUG_MODE:
+                print(f"{cls}._maybe_init({obj_or_path}) failed: {e}")
+            pass
+
+        for child_cls in cls.__subclasses__():
+            maybe_instance = child_cls._maybe_init(obj_or_path)
+            if maybe_instance is not None:
+                return maybe_instance
+
+        return None
+
+    @classmethod
+    def _tmp_path(cls: Type["SavedModel"]) -> str:
```

Comment:
Super nit: Can we add a comment here just saying that this method generates a filesystem path under our MEDIA_TMP directory every time it's called?

Reply:
added in next commit